sol/main.c

```diff
@@ -0,0 +1,507 @@
+		enqueue_reqs(sol_conf);
+		dequeue_reqs(sol_conf, tx_port_back,
+			sol_conf->tx_queue_back);
+	}
+
+	RTE_LOG(NOTICE, GATEKEEPER,
+		"sol: the Solicitor block at lcore = %u is exiting\n", lcore);
+
+	return cleanup_sol(sol_conf);
+}
+
+static int
+sol_stage1(void *arg)
+{
+	struct sol_config *sol_conf = arg;
+	int ret = get_queue_id(&sol_conf->net->back, QUEUE_TYPE_TX,
+		sol_conf->lcore_id);
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER, "sol: cannot assign a TX queue for the back interface for lcore %u\n",
+			sol_conf->lcore_id);
+		goto cleanup;
+	}
+	sol_conf->tx_queue_back = ret;
+
+	return 0;
+
+cleanup:
+	cleanup_sol(sol_conf);
+	return -1;
+}
+
+static int
+sol_stage2(void *arg)
+{
+	struct sol_config *sol_conf = arg;
+	int ret = req_queue_init(sol_conf);
+	if (ret < 0)
+		goto cleanup;
+
+	return 0;
+
+cleanup:
+	cleanup_sol(sol_conf);
+	return ret;
+}
+
+int
+run_sol(struct net_config *net_conf, struct sol_config *sol_conf)
+{
+	int ret;
+
+	if (net_conf == NULL || sol_conf == NULL) {
+		ret = -1;
+		goto out;
+	}
+
+	if (!net_conf->back_iface_enabled) {
+		RTE_LOG(ERR, GATEKEEPER, "sol: back interface is required\n");
+		ret = -1;
+		goto out;
+	}
+
+	if (sol_conf->pri_req_max_len == 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"sol: priority queue max len must be greater than 0\n");
+		ret = -1;
+		goto out;
+	}
+
+	if (sol_conf->enq_burst_size == 0 || sol_conf->deq_burst_size == 0) {
+		RTE_LOG(ERR, GATEKEEPER, "sol: priority queue enqueue and dequeue sizes must both be greater than 0\n");
+		ret = -1;
+		goto out;
+	}
+
+	if (sol_conf->deq_burst_size > sol_conf->pri_req_max_len ||
+			sol_conf->enq_burst_size > sol_conf->pri_req_max_len) {
+		RTE_LOG(ERR, GATEKEEPER, "sol: priority queue enqueue and dequeue sizes must be less than the max length of the PQ\n");
+		ret = -1;
+		goto out;
+	}
+
+	ret = init_mailbox("sol_reqs", 2 * sol_conf->pri_req_max_len,
+		sizeof(struct priority_req), sol_conf->lcore_id,
+		&sol_conf->mb);
+	if (ret < 0)
+		goto out;
+
+	ret = net_launch_at_stage1(net_conf, 0, 0, 0, 1, sol_stage1, sol_conf);
+	if (ret < 0)
+		goto out;
+
+	ret = launch_at_stage2(sol_stage2, sol_conf);
+	if (ret < 0)
+		goto stage1;
+
+	ret = launch_at_stage3("sol", sol_proc, sol_conf, sol_conf->lcore_id);
+	if (ret < 0)
+		goto stage2;
+
+	sol_conf->net = net_conf;
+
+	ret = 0;
+	goto out;
+
+stage2:
+	pop_n_at_stage2(1);
+stage1:
+	pop_n_at_stage1(1);
+out:
+	return ret;
+}
+
+/*
+ * Solicitor configuration allocated by alloc_sol_conf(), needed
+ * by gk_solicitor_enqueue() since the GK instances do not have
+ * any visibility into the Solicitor.
+ */
+static struct sol_config *sol_conf;
+
+/*
+ * There should be only one sol_config instance.
+ * Return an error if trying to allocate the second instance.
+ */
+struct sol_config *
+alloc_sol_conf(void)
+{
+	static rte_atomic16_t num_sol_conf_alloc = RTE_ATOMIC16_INIT(0);
+	if (rte_atomic16_test_and_set(&num_sol_conf_alloc) == 1) {
+		sol_conf = rte_malloc("sol_config",
```

Comment:
1. We need to bring `rte_calloc()` back here, otherwise `cleanup_sol()` may fail when it's called while Gatekeeper is still initializing. Sorry or having asked to change it to `rte_malloc()`; a comment explaining it would guarantee that nobody is going to consider the same mistake again.

2. Before returning, call `INIT_LIST_HEAD(&req_queue->head);` to guarantee that `cleanup_sol()` won't fail for the same reason above.